yarn-project/acir-simulator/src/avm/avm_context.ts

```diff
@@ -1,140 +1,179 @@
+      return results;
+    }
   }
 
   /**
-   * Create a new forked avm context - for external calls
+   * Fetch contract bytecode from world state and decode into executable instructions.
    */
-  public static newWithForkedState(executionEnvironment: AvmExecutionEnvironment, journal: AvmJournal): AvmContext {
-    const forkedState = AvmJournal.branchParent(journal);
-    return new AvmContext(executionEnvironment, forkedState);
+  private async fetchAndDecodeBytecode() {
+    // NOTE: the following is mocked as getPublicBytecode does not exist yet
+    const selector = new FunctionSelector(0);
+    const bytecode = await this.worldState.hostStorage.contractsDb.getBytecode(this.environment.address, selector);
+
+    // This assumes that we will not be able to send messages to accounts without code
+    // Pending classes and instances impl details
+    if (!bytecode) {
+      throw new NoBytecodeForContractError(this.environment.address);
+    }
+
+    this.instructions = decodeFromBytecode(bytecode);
   }
 
   /**
-   * Prepare a new AVM context that will be ready for an external call
-   * - It will fork the journal
-   * - It will set the correct execution Environment Variables for a call
+   * Prepare a new AVM context that will be ready for an external/nested call
+   * - Fork the world state journal
+   * - Derive an execution environment from the caller/parent
    *    - Alter both address and storageAddress
    *
-   * @param address - The contract to call
-   * @param executionEnvironment - The current execution environment
-   * @param journal - The current journal
+   * @param address - The contract instance to initialize a context for
+   * @param calldata - Data/arguments for nested call
+   * @param parentEnvironment - The caller/parent environment
+   * @param initialMachineState - The initial machine state (derived from call instruction args)
+   * @param parentWorldState - The caller/parent world state (journal)
    * @returns new AvmContext instance
    */
-  public static prepExternalCallContext(
+  public static async createNestedContractCallContext(
     address: AztecAddress,
     calldata: Fr[],
-    executionEnvironment: AvmExecutionEnvironment,
-    journal: AvmJournal,
-  ): AvmContext {
-    const newExecutionEnvironment = executionEnvironment.newCall(address, calldata);
-    const forkedState = AvmJournal.branchParent(journal);
-    return new AvmContext(newExecutionEnvironment, forkedState);
+    parentEnvironment: AvmExecutionEnvironment,
+    initialMachineState: InitialAvmMachineState,
+    parentWorldState: AvmWorldStateJournal,
+  ): Promise<AvmContext> {
+    const newExecutionEnvironment = parentEnvironment.deriveEnvironmentForNestedCall(address, calldata);
+    const forkedState = parentWorldState.fork();
+    const nestedContext = new AvmContext(forkedState, newExecutionEnvironment, initialMachineState);
+    await nestedContext.fetchAndDecodeBytecode();
```

Comment:
Why do you want to do the fetching and decoding here (turning this into async) if execute already does it? It feels more natural to have it in execute and then you can even avoid storing the instructions in the class, you just use them on the spot.